server/api/region.go

```diff
@@ -140,3 +141,115 @@ func (h *regionsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 	h.rd.JSON(w, http.StatusOK, regionsInfo)
 }
+
+const defaultRegionLimit = 16
+
+func (h *regionsHandler) GetTopWriteFlow(w http.ResponseWriter, r *http.Request) {
+	cluster := h.svr.GetRaftCluster()
+	if cluster == nil {
+		h.rd.JSON(w, http.StatusInternalServerError, server.ErrNotBootstrapped.Error())
+		return
+	}
+	limit := defaultRegionLimit
+	if limitStr := r.URL.Query().Get("limit"); limitStr != "" {
+		var err error
+		limit, err = strconv.Atoi(limitStr)
+		if err != nil {
+			h.rd.JSON(w, http.StatusBadRequest, err.Error())
+			return
+		}
+	}
+	regions := h.TopNRegions(cluster.GetRegions(), func(a, b *core.RegionInfo) bool { return a.WrittenBytes < b.WrittenBytes }, limit)
+	regionInfos := make([]*regionInfo, len(regions))
+	for i, r := range regions {
+		regionInfos[i] = newRegionInfo(r)
+	}
+	res := &regionsInfo{
+		Count:   len(regions),
+		Regions: regionInfos,
+	}
+	h.rd.JSON(w, http.StatusOK, res)
+}
+
+func (h *regionsHandler) GetTopReadFlow(w http.ResponseWriter, r *http.Request) {
+	cluster := h.svr.GetRaftCluster()
+	if cluster == nil {
+		h.rd.JSON(w, http.StatusInternalServerError, server.ErrNotBootstrapped.Error())
+		return
+	}
+	limit := defaultRegionLimit
+	if limitStr := r.URL.Query().Get("limit"); limitStr != "" {
+		var err error
+		limit, err = strconv.Atoi(limitStr)
+		if err != nil {
+			h.rd.JSON(w, http.StatusBadRequest, err.Error())
+			return
+		}
+	}
+	regions := h.TopNRegions(cluster.GetRegions(), func(a, b *core.RegionInfo) bool { return a.ReadBytes < b.ReadBytes }, limit)
+	regionInfos := make([]*regionInfo, len(regions))
+	for i, r := range regions {
+		regionInfos[i] = newRegionInfo(r)
+	}
+	res := &regionsInfo{
+		Count:   len(regions),
+		Regions: regionInfos,
+	}
+	h.rd.JSON(w, http.StatusOK, res)
+}
+
+// RegionHeap implements heap.Interface, used for selecting top n regions.
+type RegionHeap struct {
+	regions []*core.RegionInfo
+	less    func(a, b *core.RegionInfo) bool
+}
+
+func (h *RegionHeap) Len() int           { return len(h.regions) }
+func (h *RegionHeap) Less(i, j int) bool { return h.less(h.regions[i], h.regions[j]) }
+func (h *RegionHeap) Swap(i, j int)      { h.regions[i], h.regions[j] = h.regions[j], h.regions[i] }
+
+// Push pushes an element x onto the heap.
+func (h *RegionHeap) Push(x interface{}) {
+	h.regions = append(h.regions, x.(*core.RegionInfo))
+}
+
+// Pop removes the minimum element (according to Less) from the heap and returns
+// it.
+func (h *RegionHeap) Pop() interface{} {
+	pos := len(h.regions) - 1
```

Comment:
I think it would be better check whether Len() is zero

Reply:
Nice catch.

Reply:
It is the caller's responsibility to avoid poping an empty heap. You can refer the heap example: https://golang.org/src/container/heap/example_intheap_test.go

Reply:
got it